Added events for crop growth

Draco18s opened the pull request:
New block event: CropGrowEvent
Patches for BlockChorusFlower, BlockCocoa, BlockCrops, BlockNetherWart,
BlockReed, and BlockStem (all of the vanilla "growing" blocks).

Example mod:
https://github.com/Draco18s/ReasonableRealism/tree/Farming-Example

Draco18s committed:
Added events for crop growth

New block event
Patches for BlockChorusFlower, BlockCocoa, BlockCrops, BlockNetherWart,
BlockReed, and BlockStem (all of the vanilla "growing" blocks).

Draco18s committed:
Tabs to spaces

kashike commented on patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch:
```diff
@@ -0,0 +1,204 @@
+-                    {
+-                        this.func_185605_c(p_180650_1_, p_180650_2_);
+-                    }
++                    BlockEvent ev = new BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
```
...fix your formatting. You should not be changing the patch like this.

Draco18s replied:
I didn't touch the patch files directly.  When I fixed the formatting I ran genPatches again and sync'ed the altered patch files.

LexManos replied:
NEVER 'fix' formatting of vanilla code...

Draco18s replied:
Jesus christ, and here I thought I was trying to make sure I was using four spaces and not tabs and figured the only place there would be tabs was in front of any code _I_ had written.  I was trying to be particularly careful about avoiding altering the formatting of the vanilla code I was working around (as my IDE kept trying to "correct" things).

Draco18s replied:
Oh, and this particular section of code is wrapped in an extra indentation.  There's an additional if(){...} around the vanilla code, that's why it's tabbed in farther.  Should it not be?  The style guide makes no mention of this.

covers1624 replied:
Keep patches as small as possible, so don't indent it at all. just a brace on the same indent as the previous will be fine.

Draco18s replied:
Ok thank you. I was trying to follow the "indent your code properly" convention as I saw nothing else about it (and still don't).

liach commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -267,5 +269,54 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+         * <br>
+         */
+        public static class Pre extends CropGrowEvent {
+            public Pre(World world, BlockPos pos, IBlockState state) {
```
`@HasResult`

covers1624 commented on patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch:
```diff
@@ -0,0 +1,204 @@
+ import net.minecraft.world.World;
++import net.minecraftforge.common.MinecraftForge;
++import net.minecraftforge.event.world.BlockEvent;
++import net.minecraftforge.fml.common.eventhandler.Event;
```
Don't use imports in patches. Use fully qualified names.

Draco18s replied:
Ok, thanks, but that might want to be added to the style guide. How would I have known?

KnightMiner replied:
From [CONTRIBUTING.md](https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/CONTRIBUTING.md) (which appears at the top when making an issue/PR), there is a link leading [here](https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge#conventions-for-coding-patches-for-a-minecraft-class-javapatch) which mentions said style guideline.

Draco18s replied:
My apologies, it was late and I was frustrated and didn't see the "more info here" link.

liach commented on patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch:
```diff
@@ -0,0 +1,204 @@
+-                    {
+-                        this.func_185605_c(p_180650_1_, p_180650_2_);
+-                    }
++                    BlockEvent ev = new BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
```
and you should post the event only after random check has passed.

Draco18s committed:
Fixed styling

Draco18s commented:
Ok, I think I have addressed the issues people had with things.  I am just not sure how to squash the commits on the branch as I don't want to rebase onto the master of my fork.

iTitus commented on patches/minecraft/net/minecraft/block/BlockCactus.java.patch:
```diff
@@ -9,7 +9,41 @@
++                if(ev.getResult() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) {
++                    if (j == 15)
++                    {
++                        p_180650_1_.func_175656_a(blockpos, this.func_176223_P());
```
This seems to be the same paragraph as before, just indented.
Just leave it unindented and change only necessary lines.

Draco18s replied:
Thanks. I was sure I'd had done that (extra indentation) in another file other than BlockChorusFlower but when I was reviewing the java code, and then the generated patches, I didn't find it.

Draco18s committed:
Removed indentation

williewillus commented:
is this meant to be general enough to where mods could fire this event themselves?

If so, I would make the comments more general to match (instead of enumerating the vanilla cases)

Draco18s commented:
@williewillus Essentially yes.  I figure _most_ mod crops would be inheriting from one of those classes without needing to override (and unfortunately the logic can't be handled by BlockBush) but in the case that they don't they'd have to fire the event themselves. Suggested rephrasing?

williewillus commented:
Just don't restrict it to vanilla.

Something like:

Fired when any ~~of the~~ "growing age" blocks (**such as cacti, chorus plants, or crops in vanilla**) attempt to advance to the next growth age state during a random tick.

Draco18s commented:
Ok, thanks. Updating that now.

Draco18s committed:
Updated event description

evieluvsrainbows commented:
@Draco18s Don't worry about squashing your commits. @LexManos and the other members who have write access can automatically squash your commits via GitHub itself since GH added that ability a while ago.

Draco18s commented:
@KamranMackey I asked because it was in the contributor's readme.
Which other parts thereof are out of date, such as "how to install" being flat wrong.

LexManos commented:
The instruction for contributing to Forge have not changed in 3 years. Not sure why you've had such issues but its a simple process.

Draco18s commented:
@LexManos as mentioned in the reply to the comment where you gave me working instructions... "the current readme doesn't say that."  Current info says "gradlew setup" and your comment was "gradlew _setupForge_" followed by _____ (it literally says "you're done") vs. "import existing project."

LexManos commented:
No idea what the hell you're looking at, but the wiki is just fine: https://github.com/MinecraftForge/MinecraftForge/wiki/If-you-want-to-contribute-to-Forge Ah found where you might be looking, it still says setupForge. But the eclipse folder no longer exists. So meh.

Draco18s commented:
@LexManos The... readme on the front page of the repo...?
https://github.com/MinecraftForge/MinecraftForge/blob/1.10.x/README.md
Which incidentally also does not contain a link to the wiki for more information.

LexManos commented on patches/minecraft/net/minecraft/block/BlockCactus.java.patch:
```diff
@@ -9,7 +9,27 @@
+-
++                net.minecraftforge.event.world.BlockEvent ev = new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
++                net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(ev);
++                if(ev.getResult() != net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) {
```
This is where you would use ForgeHooks having this big block of code in each patch over and over is bad.

LexManos commented on patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch:
```diff
@@ -0,0 +1,27 @@
+-
+-                if (i < 5 && p_180650_4_.nextInt(1) == 0)
++                
++                if (i < 5)
```
First off, spacing.
Also, a 'ForgeHooks.cropsPre(x,y,z, rand.nextInt(1)==0)' would make this code and a lot of other patches A LOT cleaner.

LexManos commented on patches/minecraft/net/minecraft/block/BlockCocoa.java.patch:
```diff
@@ -1,6 +1,31 @@
++                int i = ((Integer)p_180650_3_.func_177229_b(field_176501_a)).intValue();
++
++                if (i < 2)
++                {
```
You tab this section of code in another layer which makes the patches bigger and makes it hard to tell what you add and what is vanilla, fix that.

LexManos commented on patches/minecraft/net/minecraft/block/BlockCrops.java.patch:
```diff
@@ -1,6 +1,21 @@
+                 float f = func_180672_a(this, p_180650_1_, p_180650_2_);
+-
+-                if (p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0)
++                net.minecraftforge.event.world.BlockEvent ev = new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
```
See above comment about ForgeHooks.

LexManos commented on patches/minecraft/net/minecraft/block/BlockNetherWart.java.patch:
```diff
@@ -9,7 +9,24 @@
+     {
+         int i = ((Integer)p_180650_3_.func_177229_b(field_176486_a)).intValue();
+ 
+-        if (i < 3 && p_180650_4_.nextInt(10) == 0)
```
ForgeHooks and unindent.

LexManos commented on patches/minecraft/net/minecraft/block/BlockReed.java.patch:
```diff
@@ -9,7 +9,37 @@
+                     int j = ((Integer)p_180650_3_.func_177229_b(field_176355_a)).intValue();
+-
+-                    if (j == 15)
+-                    {
```
Again nasty patch changing indent and not making it clear whats vanilla and what your changes are.

LexManos commented on patches/minecraft/net/minecraft/block/BlockStem.java.patch:
```diff
@@ -1,6 +1,18 @@
++            net.minecraftforge.event.world.BlockEvent ev = new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Pre(p_180650_1_,p_180650_2_,p_180650_3_);
++            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(ev);
++            if(ev.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW || (ev.getResult() == net.minecraftforge.fml.common.eventhandler.Event.Result.DEFAULT && p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0))
+             {
```
again so much cleaner if you did: 
if ForgeHooks.growPre(x, y, z, rand.nextInt((int)(25.0F / f) + 1) == 0)

LexManos commented on src/main/java/net/minecraftforge/event/world/BlockEvent.java:
```diff
@@ -267,5 +269,54 @@ public CreateFluidSourceEvent(World world, BlockPos pos, IBlockState state)
+     * Fired when a crop block grows.  See subevents.
+     *
+     */
+    public static class CropGrowEvent extends BlockEvent {
```
brace

Draco18s commented:
Thanks for the comments, @LexManos I'll see about fixing those up.

Draco18s committed:
Forge Hooks

Draco18s commented:
@LexManos Alright, I did the ForgeHooks thing just for BlockChorusFlower.  Is that what you are expecting? Wanted to make sure I did it right.

LexManos commented on patches/minecraft/net/minecraft/block/BlockChorusFlower.java.patch:
```diff
@@ -0,0 +1,19 @@
+                     {
+                         this.func_185605_c(p_180650_1_, p_180650_2_);
+                     }
++                    net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.BlockEvent.CropGrowEvent.Post(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_1_.func_180495_p(p_180650_2_)));
```
Could probably make this smaller with ForgeHooks to.

Draco18s replied:
Looks like it would be...shorter by 69 characters.
Do it, yes no?

Draco18s replied:
@LexManos Didn't think I would need to include an @mention. Using ForgeHooks would save 69 characters for that line.  Do you want me to do that? (I don't know what the threshold for saving size is and am asking for direction, I'd really like to get this merged so I can start using it).

LexManos replied:
Didn't think I would have to answer a stupid question that I told you to do...
Does it make the code look better, and easier to understand, does it make the code less repetitious in the other dozen patches you do?
Then yes it probably should be cleaned up.

Draco18s replied:
Thanks @LexManos I am only trying to understand your wishes more clearly.  "Could probably make this smaller" isn't a very clear instruction (the sentence contains two conditional clauses).  "If this can be made smaller by using a ForgeHook, please do so" would have been more instructive.  Thank you for your input and I will update the other classes in this manner as well.

Draco18s committed:
Simplification of patches

Draco18s committed:
A little more cleanup

Draco18s commented:
How about over here, what's holding this back?

Draco18s committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.10.x' into Crop-Growth-Events

Draco18s commented:
Hey @LexManos any additional changes needed for this to get merged?  I've had no comments for a week+

LexManos commented on patches/minecraft/net/minecraft/block/BlockCactus.java.patch:
```diff
@@ -9,7 +9,26 @@
+             if (i < 3)
+             {
+                 int j = ((Integer)p_180650_3_.func_177229_b(field_176587_a)).intValue();
+-
```
Why delete the line?

Draco18s replied:
@LexManos I am doing my best to follow the "how to contribute" guidelines.  In this case here I put the new line of code onto an existing blank line.  If that causes a patch change that is undesirable, then the how-to wiki page needs to be updated to reflect that case.  It currently has the insert of a new statement between two lines with _no_ blank line between them stating "don't add blank lines."  I used one that was there, if that was wrong, then I have been given bad instructions.

LexManos replied:
its about balancing the patch file size and making things look okay in the end code.
This for example is just dumb, there is no real reason to delete the whitepsace as it just ads extra onto the patch file.
And deleting the whitespace makes the code more of a clump and harder to read.

Making your patch file bigger just to futz with whitespace is bad, your changes should be JUST whats needed to make your changes work.

Draco18s replied:
@LexManos My point was that I was _attempting not to mess with the whitespace_ and the how to contribute page does not make it clear _how to do that_ so I made a best guess. If I did it wrong, that's fine, I'll fix it, but the instructions on what I did wrong don't exist.  I can fix this here and now, but that doesn't help someone else who's new.  Prior to starting this endeavor _I didn't know how to read patch files._  I'm slowly starting to understand what it is that you want, but please, rather than being snarky all you have to say is "this added that extra line that makes the patch files harder to read. It's cleaner if you only insert lines, even if there's already a blank one."

LexManos commented on patches/minecraft/net/minecraft/block/BlockCrops.java.patch:
```diff
@@ -1,5 +1,18 @@
+                 float f = func_180672_a(this, p_180650_1_, p_180650_2_);
+-
+-                if (p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0)
++                if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(p_180650_1_, p_180650_2_, p_180650_3_, p_180650_4_.nextInt((int)(25.0F / f) + 1) == 0))
```
change whitespace?

LexManos commented on patches/minecraft/net/minecraft/block/BlockReed.java.patch:
```diff
@@ -9,7 +9,26 @@
+                 if (i < 3)
+                 {
+                     int j = ((Integer)p_180650_3_.func_177229_b(field_176355_a)).intValue();
+-
```
...

LexManos commented on patches/minecraft/net/minecraft/block/BlockStem.java.patch:
```diff
@@ -1,6 +1,16 @@
+         if (p_180650_1_.func_175671_l(p_180650_2_.func_177984_a()) >= 9)
+         {
+             float f = BlockCrops.func_180672_a(this, p_180650_1_, p_180650_2_);
+-
```
...

Draco18s committed:
Patch cleanup

No more whitespace alterations.

Draco18s commented:
Ok, all patches cleaned of white space alterations per @LexManos instruction

LexManos commented:
Jesus you still cant figure out how to do whitespace STOP IT.
But whatever -.-
